Add Windows Support

chris-dudley opened the pull request:
Closes #11 

Adds support for reading SMBIOS data on Windows by using the [GetSystemFirmwareTable](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724379(v=vs.85).aspx) function available in kernel32.dll.

Let me know if there are any changes you would like me to make, or if you want me to explain any of the choices I made.

chris-dudley committed:
smbios: add windows support.

A copy of the raw SMBIOS table is obtained via the GetSystemFirmwareTable
function provided in kernel32.dll.
Not all of the entry point data is available, at least we have access
to all of the structures.

chris-dudley committed:
smbios: remove leftover debug log statement.

chris-dudley committed:
meta: update AUTHORS

chris-dudley committed:
readme: add notes about Windows support.

chris-dudley committed:
smbios: use native byte order in windows stream.

Should make the retrieved table size correct on big-endian
architectures.

mdlayher commented on smbios/byteorder_big.go:
```diff
@@ -0,0 +1,27 @@
+	"encoding/binary"
+)
+
+var nativeByteOrder = binary.BigEndian
```
Instead of doing a list of architectures to determine endianness, we can do it on the fly with a bit of unsafe:

```go
// nativeEndian returns the native byte order of this system.
func nativeEndian() binary.ByteOrder {
	// Determine endianness by interpreting a uint16 as a byte slice.
	v := uint16(1)
	b := *(*[2]byte)(unsafe.Pointer(&v))

	if b[0] == 1 {
		return binary.LittleEndian
	}

	return binary.BigEndian
}
```

You also made me realize that I probably need to work with native endianness on other machines too!  I assume the contents of RAM is stored in the CPU's native endianness.

chris-dudley replied:
Yeah, I've seen that way of determining endianess before. I wasn't sure how much you cared about doing tricks with unsafe, so I tried to avoid it. I'll change it over to using this method instead.

mdlayher replied:
In this kind of systems package, I'm totally cool with using unsafe.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+
+// WindowsEntryPoint contains SMBIOS Table entry point data returned from
+// GetSystemFirmwareTable. As raw access to the underlying memory is not given,
+// the full bredth of information is not available.
```
*breadth

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+	procGetSystemFirmwareTable = libKernel32.NewProc("GetSystemFirmwareTable")
+)
+
+// WindowsEntryPoint contains SMBIOS Table entry point data returned from
```
Bummer that another type is needed for Windows.  The type definition should live in a file without any build tags so that way it can be used on all platforms in type switches and such.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+func stream() (io.ReadCloser, EntryPoint, error) {
+	// Call first with empty buffer to get size
```
Nit: period so comments are complete sentences.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+	// Call first with empty buffer to get size
+	r1, _, err := procGetSystemFirmwareTable.Call(
+		uintptr(firmwareTableProviderSigRSMB),
+		uintptr(0),
```
I believe you can omit these three conversions here and just pass constant `0` and the compiler will infer the right type.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+		uintptr(0),
+	)
+
+	if r1 == 0 {
```
Just to be clear, `r1 == 0` always if `err != nil`?  Do you also need to do an explicit error check?

chris-dudley replied:
`LazyDLL.Call` always returns `err != nil` (see [here](https://golang.org/pkg/syscall/?GOOS=windows&GOARCH=amd64#LazyProc.Call)).

I have to check the primary return value (`r1`) to determine if an error occurred. For this particular call, `r1 == 0` if the call failed (as per the [MSDN documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724379(v=vs.85).aspx)).

mdlayher replied:
TIL, thanks for sharing. Mind dropping a link to that GoDoc for future readers?

chris-dudley replied:
Will do.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+	//		BYTE 	SMBIOSTableData[];
+	//	}
+
+	tableSize := nativeByteOrder.Uint32(buffer[4:8])
```
Should probably do a bounds check on this in case buffer is smaller than expected.

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+	//	}
+
+	tableSize := nativeByteOrder.Uint32(buffer[4:8])
+	// Paraoid check to make sure we don't try to go past the end of the buffer
```
*Paranoid

Not sure I understand the thought process here; can the byte order be wrong?

mdlayher commented on smbios/stream_windows.go:
```diff
@@ -0,0 +1,113 @@
+		return nil, nil, fmt.Errorf("failed to read SMBIOS data: expected %d bytes, read %d bytes: %v", bufferSize, r1, err)
+	}
+
+	// When calling GetSystemFirmwareTable with FirmwareTableProviderSignature = "RSMB",
```
From this point on, let's break out into a new function that unpacks the structure from the buffer, so it's testable.

```go
func windowsStream(buf []byte) (io.ReadCloser, EntryPoint, error) {}
```

Then please add some unit tests that ensure that the stream and entry point are parsed properly.

mdlayher reviewed (commented):
Looking pretty good overall, just would like to see some unit tests as well!

chris-dudley committed:
smbios: switch to runtime endianess check

chris-dudley committed:
smbios: improve windows stream error checking and docs

Added several checks for various possible errors, edge cases,
and sanity checks.
Added comments detailing why various checks are made with links
to relevant documentation.

chris-dudley committed:
smbios: make comments complete sentences

chris-dudley committed:
smbios: move WindowsEntryPoint struct definition.

Moved the definition into entrypoint.go so that it is available on
all platforms. This allows it to be used in type switches on all
platforms.

chris-dudley committed:
smbios: remove unnecessary integer casts

Removed unecessary uinptr(0) casts, Go will promote integer constants
automatically to fill the destination type.

Also added a bit more explanation to how the firmware table provider constant
was calculated.

chris-dudley committed:
smbios: factor out and test windows parsing

Factored out the code that parses the buffer containing the
RawSMBIOSData struct into its own function.
Added tests for this parsing to ensure that it handles various scenarios
properly.

mdlayher reviewed (approved):
LGTM, thanks very much!
